### PR TITLE
Feat/train test step

### PR DESCRIPTION
**Pull request by sanatsingh**
Implementing train and test steps

- [x] Added train loss computation
- [ ] Added eval loss computation
- [ ] Defined train and test steps

### CODE REVIEW

File: pynet/model/pynet.py [hunk 1]
```diff
@@ -132,3 +132,147 @@ def save_weights(self, filepath, overwrite=True, save_format=None, options=None)
 
     def load_weights(self, filepath, skip_mismatch=False, by_name=False, options=None):
         self.network.load_weights(filepath, skip_mismatch, by_name, options)
+
+    def compile(
+        self,
+        mse_loss: keras.losses.Loss,
+        perceptual_loss: keras.losses.Loss,
+        ssim_loss: keras.losses.Loss,
```
Comment (soumik12345) [inline pynet/model/pynet.py:140]: Nitpick: since we already know the losses being used, it's better if we define the losses directly inside `compile()` instead of passing them.

File: pynet/model/pynet.py [hunk 2]
```diff
@@ -132,3 +132,147 @@ def save_weights(self, filepath, overwrite=True, save_format=None, options=None)
...
+        l3_mse_loss = self.mean_squared_error(l3_ground_truth, l3_out_final)
+        l2_mse_loss = self.mean_squared_error(l2_ground_truth, l2_out_final)
+        l1_mse_loss = self.mean_squared_error(l1_ground_truth, l1_out_final)
+        l0_mse_loss = self.mean_squared_error(l0_ground_truth, l0_out_final)
+        l5_loss_psnr = 20 * math.log10(1.0 / math.sqrt(l5_mse_loss))
+        l4_loss_psnr = 20 * math.log10(1.0 / math.sqrt(l4_mse_loss))
+        l3_loss_psnr = 20 * math.log10(1.0 / math.sqrt(l3_mse_loss))
+        l2_loss_psnr = 20 * math.log10(1.0 / math.sqrt(l2_mse_loss))
+        l1_loss_psnr = 20 * math.log10(1.0 / math.sqrt(l1_mse_loss))
+        l0_loss_psnr = 20 * math.log10(1.0 / math.sqrt(l0_mse_loss))
```
Comment (soumik12345) [inline pynet/model/pynet.py:234]: Use [`tf.image.psnr`](https://www.tensorflow.org/api_docs/python/tf/image/psnr)

File: pynet/model/pynet.py [hunk 3]
```diff
@@ -132,3 +132,147 @@ def save_weights(self, filepath, overwrite=True, save_format=None, options=None)
...
+            ],
+            "ssim_losses": [l0_ssim, l1_ssim],
+            "perceptual_losses": [
+                l0_perceptual_loss,
+                l1_perceptual_loss,
+                l2_perceptual_loss,
+                l3_perceptual_loss,
+                l4_perceptual_loss,
+            ],
+        }
```
Comment (soumik12345) [inline pynet/model/pynet.py:272]: It should be a dictionary mapper string to metric, not list of metrics.

**Review by soumik12345 — CHANGES_REQUESTED** [review]
@sanatsingh 
Mighty work. on the PR!
Requested a few changes.

**Comment by soumik12345** [comment]
Fixes https://github.com/soumik12345/pynet/issues/6

File: pynet/model/pynet.py (hunk 2 above)
Comment (sanatsingh) [reply pynet/model/pynet.py:234]: We are already calculating MSE so reusing it for psnr calculation, although tf.image.psnr would look cleaner. LMK what you think

Comment (soumik12345) [reply pynet/model/pynet.py:235]: Using [`tf.image.psnr`](https://www.tensorflow.org/api_docs/python/tf/image/psnr) will make it more readable.

File: pynet/model/pynet.py (hunk 1 above)
Comment (soumik12345) [reply pynet/model/pynet.py]: @sanatsingh I meant that we should "define" the losses inside `compile` instead of passing them as arguments, like [this](https://github.com/keras-team/keras-core/blob/main/examples/keras_io/tensorflow/vision/zero_dce.py#L316).

File: pynet/model/pynet.py [hunk 4]
```diff
@@ -132,3 +133,146 @@ def save_weights(self, filepath, overwrite=True, save_format=None, options=None)
...
+            # SSIM Loss
+            "l0_ssim": l0_ssim,
+            "l1_ssim": l1_ssim,
+            # Perceptual Loss
+            "l0_perceptual_loss": l0_perceptual_loss,
+            "l1_perceptual_loss": l1_perceptual_loss,
+            "l2_perceptual_loss": l2_perceptual_loss,
+            "l3_perceptual_loss": l3_perceptual_loss,
+            "l4_perceptual_loss": l4_perceptual_loss,
+        }
```
Comment (soumik12345) [inline pynet/model/pynet.py:272]: I had some [conversation](https://github.com/keras-team/keras-core/issues/485) with [François](https://github.com/fchollet) recently, even though this would ensure that the values would turn up in the progress bar, it won't be available in the callback (so, no experiment tracking 🥲) or on `history.history`. So, the best way to solve this issue is to create `Mean` trackers for the different metrics to track and call `update_state()` on them in `train_step`. You can refer to [this example](https://github.com/keras-team/keras-core/blob/main/examples/keras_io/tensorflow/vision/zero_dce.py#L316).

**Review by soumik12345 — CHANGES_REQUESTED** [review]
Hi @sanatsingh 
I have left one feedback.
Let's also resolve our unresolved conversations from the previous round of feedback (this will also mend the CI).

File: pynet/model/pynet.py (hunk 2 above)
Comment (sanatsingh) [reply pynet/model/pynet.py:235]: Done

File: pynet/model/pynet.py (hunk 1 above)
Comment (sanatsingh) [reply pynet/model/pynet.py]: Ok, I defined them as default values so as to allow change in loss functions

Comment (sanatsingh) [reply pynet/model/pynet.py]: Done

**Review by sanatsingh — COMMENTED** [review]
Changes made

File: pynet/model/pynet.py (hunk 4 above)
Comment (sanatsingh) [reply pynet/model/pynet.py]: Done

**Review by soumik12345 — APPROVED** [review]
Thanks for the PR!
